File: vuln-sandbox/php/cwe-89/sql-injection-safe.php
<?php
// CWE-89: SQL Injection - Safe Pattern
// Rule: php-sql-string-concat
// Expected: TrueNegative

$pdo = new PDO("mysql:host=localhost;dbname=db", "user", "pass");

// SAFE: Using PDO prepared statements with bound parameters
$id = $_GET['id'];
$stmt = $pdo->prepare("SELECT * FROM users WHERE id = ?");
$stmt->execute([$id]);
$result = $stmt->fetchAll();

// SAFE: Named parameters
$username = $_POST['username'];
$stmt2 = $pdo->prepare("SELECT * FROM accounts WHERE username = :username");
$stmt2->execute([':username' => $username]);
?>

File: vuln-sandbox/java/cwe-89/SqlInjectionSafe.java
// SAFE: java-sqli-statement-execute-concat — PreparedStatement prevents SQL injection
// Rule: java-sqli-statement-execute-concat | CWE-89 | Expected: TrueNegative

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class SqlInjectionSafe extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) {
        String userId = request.getParameter("id");

        try {
            Connection conn = DriverManager.getConnection("jdbc:mysql://localhost/app", "root", "");

            // SAFE: PreparedStatement uses a parameterized placeholder; user input never concatenated
            PreparedStatement stmt = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
            stmt.setString(1, userId);
            ResultSet rs = stmt.executeQuery();

            while (rs.next()) {
                response.getWriter().println(rs.getString("username"));
            }

            conn.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
